MM-1584 Private message channels no longer do a page refresh on creation/first use

rgarmsen2295 opened the pull request:
I believe this should also fix the issue with online status icons not appearing for uncreated PM channels as a side-effect.  When a user clicks on a new PM channel instead of refreshing, a new client call is made to get the server to create the channel that causes the user to switch to the PM channel on completion.  The online status icon is replaced by a loading gif while this occurs, which is usually less than a second except on slow connections.  On failure it simply falls back to the original method of channel creation.

coreyhulen commented on web/react/components/more_direct_channels.jsx:
```diff
@@ -3,67 +3,107 @@
+                    e.preventDefault();
+                    self.setState({loadingPMChannel: index});
+
+                    Client.createPMChannelIfNotExists(channel, otherUserId,
```
minor nit, but why not createDirectChannel().  I think the backend refers to these as direct channels vs private channels (which is labeled a private group in the UI).

there is:
public channel
private channel
direct channel (1:1)

rgarmsen2295 replied:
Sure, was 0/5 on it to begin with, wasn't sure which to go with originally (since the client functions straddle between front and backend)

coreyhulen commented on web/react/components/sidebar.jsx:
```diff
@@ -242,17 +245,17 @@ module.exports = React.createClass({
-                var teammate_username = utils.getDirectTeammate(channel.id).username;
-                document.title = teammate_username + ' ' + document.title.substring(document.title.lastIndexOf('-'));
+                var teammateUsername = utils.getDirectTeammate(channel.id).username;
+                document.title = teammateUsername + ' ' + document.title.substring(document.title.lastIndexOf('-'));
```
maybe use module.exports.getDisplayName?

rgarmsen2295 replied:
I think we are deliberately not using nicknames/fullnames for PM channels currently, as per #322

coreyhulen commented on web/react/components/sidebar.jsx:
```diff
@@ -347,38 +356,63 @@ module.exports = React.createClass({
             } else {
-                href = TeamStore.getCurrentTeamUrl() + '/channels/' + channel.name;
+                // It's a direct message channel that doesn't exist yet so let's create it now
+                var ids = channel.name.split('__');
```
duplicate code?  should we refactor.

rgarmsen2295 replied:
Yup, I'll make a function to handle getting the other user id from a PM channel name

rgarmsen2295 commented:
Okay made some changes as per your feedback @coreyhulen please take another look when you get a chance, thanks!

coreyhulen commented on web/react/components/more_direct_channels.jsx:
```diff
@@ -3,67 +3,102 @@
     },
     getInitialState: function() {
-        return { channels: [] };
+        return {channels: [], loadingPMChannel: -1};
```
I still see refs to PMChannel

rgarmsen2295 replied:
Oh so remove _all_ references to private message channels.  Okay, misunderstood, thought it was only stuff dealing with the backend since we call direct message channels private message channels to users, but it does make more sense to be completely consistent with variables and functions one way or the other, I'll change it when I get a chance tomorrow.

coreyhulen commented on web/react/components/sidebar.jsx:
```diff
@@ -347,38 +356,57 @@ module.exports = React.createClass({
+                        );
+                    };
+                }
             }
```
sorry meant, all these lines are dup? not just the 1.

rgarmsen2295 replied:
There are slight differences that would require separate if-then paths (in the "more" modal the success function includes a call to hide the modal for example) but I could refactor it further.  I was hesitant to do anymore given that I'm doing things like setting the state of the object, etc. which I don't think we ever do outside of the main component file (at least it's never called in any utils function).

coreyhulen replied:
ok...don't bother then.  I didn't realize they are slightly different.

jwilander commented:
This needs a rebase.

rgarmsen2295 commented:
Okay rebase complete and changed references to PM channels over to DM channels for the stuff I added @coreyhulen please take another look when you get a moment, thanks!

coreyhulen commented:
+1 looks good

jwilander commented:
+1
